File: src/main/java/com/goudourasv/data/lectures/LecturesRepository.java
package com.goudourasv.data.lectures;

import com.goudourasv.data.courses.CourseEntity;
import com.goudourasv.domain.lectures.Lecture;
import com.goudourasv.http.lectures.dto.LectureCreate;
import com.goudourasv.http.lectures.dto.LectureUpdate;

import javax.enterprise.context.ApplicationScoped;
import javax.persistence.EntityManager;
import javax.persistence.Query;
import javax.ws.rs.NotFoundException;
import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static com.goudourasv.data.lectures.LecturesMapper.toLecture;
import static com.goudourasv.data.lectures.LecturesMapper.toLectures;

@ApplicationScoped
public class LecturesRepository {
    private final EntityManager entityManager;

    public LecturesRepository(EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    public List<Lecture> getFilteredLectures(UUID courseId) {
        String sqlQuery = "SELECT * FROM lectures";
        Map<String, Object> parametersMap = new HashMap<>();

        if (courseId != null) {
            sqlQuery += " WHERE course_id = :courseId";
            parametersMap.put("courseId", courseId);
        }

        Query query = entityManager.createNativeQuery(sqlQuery, LectureEntity.class);
        for (String key : parametersMap.keySet()) {
            query.setParameter(key, parametersMap.get(key));
        }

        @SuppressWarnings("unchecked")
        List<LectureEntity> lectureEntities = query.getResultList();
        List<Lecture> lectures = toLectures(lectureEntities);

        return lectures;
    }

    public Lecture getSpecificLecture(UUID lectureId) {
        LectureEntity lectureEntity = entityManager.find(LectureEntity.class, lectureId);
        CourseEntity courseEntity = lectureEntity.getCourseEntity();
        Lecture lecture = toLecture(lectureEntity, courseEntity);
        return lecture;
    }

    public Lecture createLecture(LectureCreate lectureCreate) {
        LectureEntity lectureEntity = new LectureEntity();
        lectureEntity.setTitle(lectureCreate.getTitle());
        lectureEntity.setStartTime(lectureCreate.getStartTime());
        lectureEntity.setEndTime(lectureCreate.getEndTime());

        CourseEntity courseEntity = entityManager.getReference(CourseEntity.class, lectureCreate.getCourseId());
        lectureEntity.setCourseEntity(courseEntity);
        entityManager.persist(lectureEntity);
        entityManager.flush();

        Lecture lecture = toLecture(lectureEntity, courseEntity);
        return lecture;

    }

    public boolean deleteSpecificLecture(UUID id) {
        LectureEntity lectureEntity = entityManager.getReference(LectureEntity.class, id);
        if (lectureEntity == null) {
            return false;
        }
        entityManager.remove(lectureEntity);
        return true;
    }

    public Lecture replaceLecture(UUID id, LectureCreate lectureCreate) {
        LectureEntity lectureEntity = new LectureEntity();
        lectureEntity.setTitle(lectureCreate.getTitle());
        lectureEntity.setStartTime(lectureCreate.getStartTime());
        lectureEntity.setEndTime(lectureCreate.getEndTime());
        lectureEntity.setId(id);

        CourseEntity courseEntity = entityManager.getReference(CourseEntity.class, lectureCreate.getCourseId());
        lectureEntity.setCourseEntity(courseEntity);
        try {
            entityManager.merge(lectureEntity);
        } catch (Exception ex) {
            throw new NotFoundException("Lecture with id: " + id + "doesn't exist");
        }
        Lecture lecture = toLecture(lectureEntity, courseEntity);
        return lecture;
    }

    public Lecture partiallyUpdateLecture(UUID id, LectureUpdate lectureUpdate) {
        LectureEntity lectureEntity = entityManager.getReference(LectureEntity.class, id);

        if (lectureUpdate.getTitle() != null) {
            String newLectureTitle = lectureUpdate.getTitle();
            lectureEntity.setTitle(newLectureTitle);
        }
        if (lectureUpdate.getStartTime() != null) {
            Instant newLectureStartTime = lectureUpdate.getStartTime();
            lectureEntity.setStartTime(newLectureStartTime);
        }
        if (lectureUpdate.getEndTime() != null) {
            Instant newLectureEndTime = lectureUpdate.getEndTime();
            lectureEntity.setEndTime(newLectureEndTime);
        }

        entityManager.merge(lectureEntity);
        entityManager.flush();
        CourseEntity courseEntity = lectureEntity.getCourseEntity();

        Lecture lecture = toLecture(lectureEntity, courseEntity);
        return lecture;


    }
}

File: src/main/java/com/goudourasv/data/lectures/LecturesRepository.kt
package com.goudourasv.data.lectures

import com.goudourasv.data.courses.CourseEntity
import com.goudourasv.domain.lectures.Lecture
import com.goudourasv.domain.lectures.LectureData
import com.goudourasv.http.lectures.dto.LectureCreate
import com.goudourasv.http.lectures.dto.LectureUpdate
import java.util.*
import javax.enterprise.context.ApplicationScoped
import javax.persistence.EntityManager
import javax.persistence.Query

@ApplicationScoped
class LecturesRepository(private val entityManager: EntityManager) {

    fun getFilteredLectures(courseId: UUID): List<Lecture> {
        var sqlQuery = "SELECT * FROM lectures"
        val parametersMap = mutableMapOf<String, Any>()
        sqlQuery += "WHERE course_id = :courseId"
        parametersMap["courseId"] = courseId

        val query: Query = entityManager.createNativeQuery(sqlQuery, LectureEntity::class.java)
        for (key in parametersMap.keys) {
            query.setParameter(key, parametersMap[key])
        }
        @Suppress("UNCHECKED_CAST")
        val lectureEntities: List<LectureEntity> = query.resultList as List<LectureEntity>
        return lectureEntities.toLectures()
    }

    fun getSpecificLecture(id: UUID): Lecture {
        val lectureEntity = entityManager.find(LectureEntity::class.java, id)
        return lectureEntity.toLecture()
    }

    fun createLecture(lectureCreate: LectureCreate): Lecture {
        val courseEntity = entityManager.getReference(CourseEntity::class.java, lectureCreate.courseId)
        val lectureEntity = LectureEntity(
            title = lectureCreate.title,
            startTime = lectureCreate.startTime,
            endTime = lectureCreate.endTime,
            uri = lectureCreate.uri,
            courseEntity = courseEntity
        )
        entityManager.persist(lectureEntity)
        entityManager.flush()

        return lectureEntity.toLecture()
    }

    fun deleteSpecificLecture(id: UUID): Boolean {
        val lectureEntity = entityManager.getReference(LectureEntity::class.java, id) ?: return false
        entityManager.remove(lectureEntity)
        return true
    }

    fun replaceLecture(id: UUID, lectureCreate: LectureCreate): Lecture {
        val courseEntity = entityManager.getReference(CourseEntity::class.java, lectureCreate.courseId)
        val lectureEntity = LectureEntity(
            id = id,
            title = lectureCreate.title,
            startTime = lectureCreate.startTime,
            endTime = lectureCreate.endTime,
            uri = lectureCreate.uri,
            courseEntity = courseEntity
        )

        entityManager.merge(lectureEntity)
        entityManager.flush()

        return lectureEntity.toLecture()
    }

    fun partiallyUpdateLecture(id: UUID, lectureUpdate: LectureUpdate): Lecture {
        val lectureEntity = entityManager.getReference(LectureEntity::class.java, id)

        if (lectureUpdate.title != null) {
            lectureEntity.title = lectureUpdate.title!!
        }

        if (lectureUpdate.startTime != null) {
            lectureEntity.startTime = lectureUpdate.startTime!!
        }

        if (lectureUpdate.endTime != null) {
            lectureEntity.endTime = lectureUpdate.endTime!!
        }

        if (lectureUpdate.uri != null) {
            lectureEntity.uri = lectureUpdate.uri!!
        }

        entityManager.merge(lectureEntity)
        entityManager.flush()

        return lectureEntity.toLecture()

    }
}
